### ISSUE TITLE
udhcpc client doesn't maintain leases

### ISSUE THREAD

**Issue by GeoffWilliams** [jpetazzo/pipework#181]
## Overview
If using `udhcpc` with pipework and docker, leases are not maintained because `udhcpc` was invoked with the `-q` option which causes it to exit after obtaining a lease.  This can cause containers to be unreachable by hostname once the lease runs out if your using a dynamic DNS server.

## Workaround
Changing the pipework script at line 324 to read:
```shell
ip netns exec "$NSPID" "$DHCP_CLIENT" -i "$CONTAINER_IFNAME" \
```
Fixes this problem and multiple `udhcpc` processes are visible in `ps` once this change has been made.

## Long-term fix
I'm not sure the above workaround is an elegant enough fix to this problem, I believe the `-q` option was added to avoid having a bunch of `udhcpc` clients hanging around trying to obtain IP addresses for long-dead containers.

In my case, the containers I want to give DHCP addresses to are ones that I want to be running all the time but this might not be the case with other peoples systems.

I'm unclear on the best way to fix this - perhaps an option to the script to skip the `-q` flag to `udhcpc` and the `kill` operation on `dhclient` would be the way to go?  I also noticed a `-q` being set for `dhcpcd` but on my debian box this flag is undefined.

Happy to have a go at putting together a PR to implement if this is acceptable, otherwise hope this information can help out anyone else struggling with this.

**Comment by Dmitriusan** [comment]
In my deployment, I'm using unique macs for containers, large private address block, and long-term (5-year) leases  on dhcp server. Of course, it is an ugly hack, but that is the way to live with that and be able to safely suspend/restart containers :)

**Comment by jpetazzo** [comment]
Indeed, I was using `-q` to avoid lingering DHCP clients; but it means that leases don't get renewed.

Another possibility would be to keep the DHCP client up and running, and then:
- either clean them up manually;
- or add a watcher process, so that the DHCP client detects if it's the only one using that network namespace, and goes away if that's the case.

**Comment by deitch** [comment]
Isn't this the same issue with dhclient, since https://github.com/jpetazzo/pipework/blob/master/pipework#L353 kills the `dhclient`, thus preventing renewal?

In a general sense, since each of these maintains their own leases database (e.g. dhclient in `/var/lib/dhclient/dhclient.leases`), what happens when I run 2 containers, and `dhclient` for each? Will they trounce each other? Why not?

**Comment by deitch** [comment]
@jpetazzo how about an option to keep dhcp alive after? It is easier to do with the `dhcp` option, since the default for udhcpc is to background itself, so you can just add `-f` to DHCP_OPTIONS. It is harder for dhclient and udhcpc since pipework expressly has the option `-q` or `kill <pid>` on it. It would require a command-line option. 

I think it is reasonable. I would expect if someone wants to run pipework in some kind of wrapper, and then take responsibility for managing (and killing) dhcp clients, let it be their issue?

**Comment by deitch** [comment]
I created a PR that allows a user to override the elimination of the dhcp process. If you call it with `dhclient`, then it works like now. But if you call it with `dhclient-f`, then it does `dhclient` remaining. A process outside pipework has to take responsibility for cleaning up processes later so they don't hang around (e.g. the pipework container).

See https://github.com/jpetazzo/pipework/pull/186

**Comment by jpetazzo** [comment]
This is addressed by #186. Thanks a lot @deitch for the contribution!

**Comment by deitch** [comment]
Pleasure. We all help each other. OK, it is your job (you work at Docker), and for me just a side-effect of a client gig, but still, all good! :-)

Your blog articles are interesting. Gotta follow you on twitter.
